# Key Questions to Answer

## Exploratory Analysis

### What’s the default rate in the dataset?

### Which features correlate most with loan defaults?

### How does interest_rate affect default risk?


In [4]:
import pandas as pd
import numpy as np

In [5]:
df=pd.read_csv('loan_data.csv')

In [6]:
df.default.value_counts(normalize=True).mul(100)

default
0    84.93
1    15.07
Name: proportion, dtype: float64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   applicant_id              10000 non-null  object 
 1   age                       10000 non-null  int64  
 2   mothly_income             10000 non-null  float64
 3   investment_income         10000 non-null  float64
 4   savings_account_balance   10000 non-null  float64
 5   checking_account_balance  10000 non-null  float64
 6   credit_card_balance       10000 non-null  float64
 7   other_loans_balance       10000 non-null  float64
 8   loan_amount_requested     10000 non-null  float64
 9   installment               10000 non-null  float64
 10  loan_date                 10000 non-null  object 
 11  loan_purpose              10000 non-null  object 
 12  state                     10000 non-null  object 
 13  city                      10000 non-null  object 
 14  credit_

In [8]:
df.default.value_counts(normalize=True).mul(100)

default
0    84.93
1    15.07
Name: proportion, dtype: float64

* Default rating is 15,07%

In [10]:
df.corr(method='spearman',numeric_only=True)

,age,mothly_income,investment_income,savings_account_balance,checking_account_balance,credit_card_balance,other_loans_balance,loan_amount_requested,installment,credit_score,loan_amount,interest_rate,loan_term,employment_duration,dependents,years_at_residence,bank_account_age,previous_defaults,default
age,1.000000,0.007207,0.002749,0.000973,0.000273,0.010376,-0.007467,-0.004517,-0.005670,-0.002902,0.005720,0.009246,0.008335,0.001214,0.003758,-0.009055,0.013341,-0.002923,0.017874
mothly_income,0.007207,1.000000,-0.010456,-0.007644,0.005669,-0.004761,-0.013096,0.000131,-0.004213,-0.013877,0.007755,0.012083,0.001283,0.000584,0.004974,0.008210,-0.001839,0.005498,-0.009904
investment_income,0.002749,-0.010456,1.000000,-0.014308,-0.018088,-0.006644,0.004907,-0.010325,0.018287,-0.016277,-0.011393,0.004871,0.009441,-0.013985,0.008647,0.000752,-0.001671,0.006904,-0.006461
savings_account_balance,0.000973,-0.007644,-0.014308,1.000000,0.008730,0.001594,-0.004626,-0.002480,-0.011306,-0.005508,-0.001614,0.001372,-0.012554,0.007396,0.004439,0.000873,0.007763,0.001623,0.013718
checking_account_balance,0.000273,0.005669,-0.018088,0.008730,1.000000,-0.014669,-0.000660,-0.017130,-0.000260,-0.004657,-0.012856,-0.013048,0.001055,-0.006788,0.000816,0.006421,-0.003902,-0.001368,0.004226
credit_card_balance,0.010376,-0.004761,-0.006644,0.001594,-0.014669,1.000000,0.004267,0.014059,-0.001102,0.001674,0.000628,-0.016578,-0.018525,-0.006574,0.005242,0.013073,-0.001119,0.000127,-0.023535
other_loans_balance,-0.007467,-0.013096,0.004907,-0.004626,-0.000660,0.004267,1.000000,-0.008421,0.008307,0.013158,0.002534,-0.004391,0.007477,-0.004926,-0.000608,-0.009948,-0.013659,-0.007622,-0.002986
loan_amount_requested,-0.004517,0.000131,-0.010325,-0.002480,-0.017130,0.014059,-0.008421,1.000000,-0.014627,-0.012768,-0.001866,-0.003198,0.004173,0.003079,-0.001431,-0.011306,0.010197,0.006411,-0.005825
installment,-0.005670,-0.004213,0.018287,-0.011306,-0.000260,-0.001102,0.008307,-0.014627,1.000000,-0.004711,-0.003958,-0.006617,0.016612,-0.002820,0.005816,0.000027,0.012758,-0.007479,-0.006143
credit_score,-0.002902,-0.013877,-0.016277,-0.005508,-0.004657,0.001674,0.013158,-0.012768,-0.004711,1.000000,-0.000605,-0.001235,-0.002984,0.015053,0.009886,-0.000391,0.003247,-0.003581,0.006120


* Credit Card and Savings Account Balance, although, it correlated pourly.

* Interest Rate affect positively the defaul measure. Higher the interest, higher the probability of default.

## Preprocessing:

### How to handle missing values in credit_score?

In [15]:
df.loc[df['default']==1,'credit_score'] = df['credit_score']*0.30

* As I created this Data set,i decided to low credit_score with defaul to 30% from what it was

In [17]:
df.previous_defaults.value_counts()

previous_defaults
1    3401
0    3328
2    3271
Name: count, dtype: int64

In [18]:
df.loc[df['previous_defaults']==1,'credit_score'] = df['credit_score']*0.80

In [19]:
df.loc[df['previous_defaults']==2,'credit_score'] = df['credit_score']*0.50

* Did almost the same with previous default,cause it impact on the credit_score (using mu knowledge).
* PS: I'm only doing this cause I created this Data Base
* Now, I can fill the missing values of the credit Score, based on the mean times percents I used with default

In [21]:
df.isna().sum()

applicant_id                  0
age                           0
mothly_income                 0
investment_income             0
savings_account_balance       0
checking_account_balance      0
credit_card_balance           0
other_loans_balance           0
loan_amount_requested         0
installment                   0
loan_date                     0
loan_purpose                  0
state                         0
city                          0
credit_score                100
loan_amount                   0
interest_rate                 0
loan_term                     0
employment_status             0
employment_duration           0
marital_status                0
dependents                    0
education_level               0
residential_status            0
years_at_residence            0
bank_account_age              0
previous_defaults             0
default                       0
dtype: int64

In [22]:
df.loc[(df['default'] == 1) & (df['credit_score'].isna()), 'credit_score'] = df['credit_score'].mean() * 0.3

In [23]:
df.loc[(df['previous_defaults'] == 1) & (df['credit_score'].isna()), 'credit_score'] = df['credit_score'].mean() * 0.8

In [24]:
df.loc[(df['previous_defaults'] == 2) & (df['credit_score'].isna()), 'credit_score'] = df['credit_score'].mean() * 0.5

In [25]:
df.loc[ (df['credit_score'].isna()), 'credit_score'] = df['credit_score'].mean()

In [26]:
df.isna().sum()

applicant_id                0
age                         0
mothly_income               0
investment_income           0
savings_account_balance     0
checking_account_balance    0
credit_card_balance         0
other_loans_balance         0
loan_amount_requested       0
installment                 0
loan_date                   0
loan_purpose                0
state                       0
city                        0
credit_score                0
loan_amount                 0
interest_rate               0
loan_term                   0
employment_status           0
employment_duration         0
marital_status              0
dependents                  0
education_level             0
residential_status          0
years_at_residence          0
bank_account_age            0
previous_defaults           0
default                     0
dtype: int64

* OK, treated the credit_score, trying to keep it close to real world

* The loan amount requested should be at least equal or higher than loan amount

In [29]:
df.loc[(df['default']) ==0 & (df['previous_defaults']==0),'loan_amount_requested']=df['loan_amount']

In [30]:
df.loc[(df['loan_amount_requested']) < (df['loan_amount']),'loan_amount'].count()

1363

In [31]:
df.loc[(df['default']) ==1 & (df['previous_defaults']>0),'loan_amount_requested']=df['loan_amount']*1.5

In [32]:
df.loc[(df['loan_amount_requested']) < (df['loan_amount']),'loan_amount'].count()

445

In [33]:
df.loc[(df['default']) ==1 & (df['previous_defaults']<=2),'loan_amount_requested']=df['loan_amount']*1.4

In [34]:
df.loc[(df['loan_amount_requested']) < (df['loan_amount']),('loan_amount','loan_amount_requested','previous_defaults','default')]

,loan_amount,loan_amount_requested,previous_defaults,default


* the monthly income, should be higher than the installment, and the installment should be loan amount,times interests, divided by loan term
* Let's adjust installment

In [36]:
df['installment'] = df.apply(
    lambda row: row['loan_amount'] * 
    (row['interest_rate'] * (1 + row['interest_rate'])**row['loan_term']) / 
    ((1 + row['interest_rate'])**row['loan_term'] - 1),
    axis=1
)

In [37]:
df['installment'] = round(df['installment'],2)

In [38]:
df.loc[((df['mothly_income']+df['investment_income'])) > (df['installment']),'mothly_income'] = df['mothly_income']

In [39]:
df.loc[((df['mothly_income']+df['investment_income'])) < (df['installment']),'mothly_income'] = df['mothly_income']*6.5

In [40]:
df.loc[((df['mothly_income'])) < (df['installment']),'default'] = 1

In [41]:
df.loc[((df['mothly_income']+df['investment_income'])) > (df['installment']),'default'] = 0

In [42]:
df['default'].value_counts(normalize=True).mul(100)

default
0    82.35
1    17.65
Name: proportion, dtype: float64

* Ok, now we got a little more close to reality

In [44]:
df.corr(numeric_only=True)

,age,mothly_income,investment_income,savings_account_balance,checking_account_balance,credit_card_balance,other_loans_balance,loan_amount_requested,installment,credit_score,loan_amount,interest_rate,loan_term,employment_duration,dependents,years_at_residence,bank_account_age,previous_defaults,default
age,1.000000,-0.000072,0.002884,0.000810,0.000305,0.010262,-0.007475,0.004514,0.009645,-0.008526,0.005659,0.009213,0.008051,0.001138,0.003777,-0.009229,0.013507,-0.002941,0.003218
mothly_income,-0.000072,1.000000,-0.058459,0.004381,-0.002927,-0.007100,-0.015464,0.447225,0.470405,-0.002344,0.481775,0.193616,-0.062653,0.005349,-0.006582,0.014059,0.007050,-0.002625,-0.258633
investment_income,0.002884,-0.058459,1.000000,-0.014106,-0.017938,-0.006618,0.004844,-0.011673,-0.002218,-0.006280,-0.011515,0.005098,0.009440,-0.014050,0.008633,0.000633,-0.001747,0.006928,-0.017767
savings_account_balance,0.000810,0.004381,-0.014106,1.000000,0.008878,0.001594,-0.004581,0.002513,-0.001537,-0.009501,-0.001518,0.001407,-0.012406,0.007354,0.004453,0.000824,0.007749,0.001650,0.007050
checking_account_balance,0.000305,-0.002927,-0.017938,0.008878,1.000000,-0.014790,-0.000722,-0.009619,-0.018536,-0.002123,-0.012731,-0.013100,0.001052,-0.006699,0.000728,0.006420,-0.003830,-0.001360,-0.012990
credit_card_balance,0.010262,-0.007100,-0.006618,0.001594,-0.014790,1.000000,0.004351,-0.005448,-0.002924,0.009239,0.000491,-0.016689,-0.018594,-0.006335,0.005272,0.013048,-0.001233,0.000074,0.005009
other_loans_balance,-0.007475,-0.015464,0.004844,-0.004581,-0.000722,0.004351,1.000000,0.003351,0.003079,0.013624,0.002616,-0.004390,0.007466,-0.004849,-0.000513,-0.010094,-0.013708,-0.007676,0.009942
loan_amount_requested,0.004514,0.447225,-0.011673,0.002513,-0.009619,-0.005448,0.003351,1.000000,0.749593,0.059091,0.930986,0.003174,-0.010332,-0.007854,0.013363,0.005531,0.018207,-0.235037,0.337403
installment,0.009645,0.470405,-0.002218,-0.001537,-0.018536,-0.002924,0.003079,0.749593,1.000000,-0.000858,0.802554,0.483367,-0.162869,-0.000677,0.003542,-0.001221,0.009199,-0.005715,0.466413
credit_score,-0.008526,-0.002344,-0.006280,-0.009501,-0.002123,0.009239,0.013624,0.059091,-0.000858,1.000000,-0.005207,-0.000769,-0.001448,0.006037,0.005176,0.004047,-0.000158,-0.513842,-0.003414


* another adjust, is that if the loan is for a house, then the type must be mortgage

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   applicant_id              10000 non-null  object 
 1   age                       10000 non-null  int64  
 2   mothly_income             10000 non-null  float64
 3   investment_income         10000 non-null  float64
 4   savings_account_balance   10000 non-null  float64
 5   checking_account_balance  10000 non-null  float64
 6   credit_card_balance       10000 non-null  float64
 7   other_loans_balance       10000 non-null  float64
 8   loan_amount_requested     10000 non-null  float64
 9   installment               10000 non-null  float64
 10  loan_date                 10000 non-null  object 
 11  loan_purpose              10000 non-null  object 
 12  state                     10000 non-null  object 
 13  city                      10000 non-null  object 
 14  credit_

In [47]:
df.loan_purpose.value_counts()

loan_purpose
Personal     2514
Education    2510
Home         2492
Car          2484
Name: count, dtype: int64

In [48]:
df.residential_status.value_counts()


residential_status
Owned        5056
Rented       3975
Mortgaged     969
Name: count, dtype: int64

In [49]:
df.loc[(df['loan_purpose'] == 'Home'),'residential_status'] = 'Mortgaged'

In [50]:
df.residential_status.value_counts()

residential_status
Owned        3800
Mortgaged    3211
Rented       2989
Name: count, dtype: int64

* If employment is unemployed, the mothly income should be zero and employment duration should be zero either

In [52]:
df.employment_status.value_counts()

employment_status
Employed         7052
Self-Employed    1944
Unemployed       1004
Name: count, dtype: int64

In [53]:
df.loc[(df['employment_status'] == 'Unemployed'),'mothly_income']= 0

In [54]:
df.loc[(df['employment_status'] == 'Unemployed'),'employment_duration']= 0

In [55]:
df.employment_status.value_counts()

employment_status
Employed         7052
Self-Employed    1944
Unemployed       1004
Name: count, dtype: int64

* that adjusted, we can go on!

In [57]:
df.corr(numeric_only=True)

,age,mothly_income,investment_income,savings_account_balance,checking_account_balance,credit_card_balance,other_loans_balance,loan_amount_requested,installment,credit_score,loan_amount,interest_rate,loan_term,employment_duration,dependents,years_at_residence,bank_account_age,previous_defaults,default
age,1.000000,0.003563,0.002884,0.000810,0.000305,0.010262,-0.007475,0.004514,0.009645,-0.008526,0.005659,0.009213,0.008051,0.006714,0.003777,-0.009229,0.013507,-0.002941,0.003218
mothly_income,0.003563,1.000000,-0.048570,0.004436,-0.003776,-0.006931,-0.022279,0.378705,0.399834,0.006678,0.408036,0.168514,-0.058573,0.196735,-0.001695,0.009459,0.009250,-0.007205,-0.231661
investment_income,0.002884,-0.048570,1.000000,-0.014106,-0.017938,-0.006618,0.004844,-0.011673,-0.002218,-0.006280,-0.011515,0.005098,0.009440,-0.002546,0.008633,0.000633,-0.001747,0.006928,-0.017767
savings_account_balance,0.000810,0.004436,-0.014106,1.000000,0.008878,0.001594,-0.004581,0.002513,-0.001537,-0.009501,-0.001518,0.001407,-0.012406,0.005483,0.004453,0.000824,0.007749,0.001650,0.007050
checking_account_balance,0.000305,-0.003776,-0.017938,0.008878,1.000000,-0.014790,-0.000722,-0.009619,-0.018536,-0.002123,-0.012731,-0.013100,0.001052,-0.004794,0.000728,0.006420,-0.003830,-0.001360,-0.012990
credit_card_balance,0.010262,-0.006931,-0.006618,0.001594,-0.014790,1.000000,0.004351,-0.005448,-0.002924,0.009239,0.000491,-0.016689,-0.018594,-0.006754,0.005272,0.013048,-0.001233,0.000074,0.005009
other_loans_balance,-0.007475,-0.022279,0.004844,-0.004581,-0.000722,0.004351,1.000000,0.003351,0.003079,0.013624,0.002616,-0.004390,0.007466,-0.010139,-0.000513,-0.010094,-0.013708,-0.007676,0.009942
loan_amount_requested,0.004514,0.378705,-0.011673,0.002513,-0.009619,-0.005448,0.003351,1.000000,0.749593,0.059091,0.930986,0.003174,-0.010332,-0.011841,0.013363,0.005531,0.018207,-0.235037,0.337403
installment,0.009645,0.399834,-0.002218,-0.001537,-0.018536,-0.002924,0.003079,0.749593,1.000000,-0.000858,0.802554,0.483367,-0.162869,-0.005209,0.003542,-0.001221,0.009199,-0.005715,0.466413
credit_score,-0.008526,0.006678,-0.006280,-0.009501,-0.002123,0.009239,0.013624,0.059091,-0.000858,1.000000,-0.005207,-0.000769,-0.001448,0.009463,0.005176,0.004047,-0.000158,-0.513842,-0.003414


In [58]:
df.city.value_counts()

city
Lake Michael            15
West Michael            13
South Michael           13
New David               11
Port Michael            11
                        ..
East Michelleborough     1
Delgadoborough           1
South Tracychester       1
Dakotaborough            1
Samuelview               1
Name: count, Length: 7694, dtype: int64

* There's too many city to convert do dummies, so, I won't be taking this column to the model.

In [60]:
df.drop(columns={'city','applicant_id'},inplace=True)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       10000 non-null  int64  
 1   mothly_income             10000 non-null  float64
 2   investment_income         10000 non-null  float64
 3   savings_account_balance   10000 non-null  float64
 4   checking_account_balance  10000 non-null  float64
 5   credit_card_balance       10000 non-null  float64
 6   other_loans_balance       10000 non-null  float64
 7   loan_amount_requested     10000 non-null  float64
 8   installment               10000 non-null  float64
 9   loan_date                 10000 non-null  object 
 10  loan_purpose              10000 non-null  object 
 11  state                     10000 non-null  object 
 12  credit_score              10000 non-null  float64
 13  loan_amount               10000 non-null  float64
 14  interes

* Now we got 7 categorical columns that we are gonna change into dummies values.

In [63]:
dfmodel = df

In [64]:
dfmodel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       10000 non-null  int64  
 1   mothly_income             10000 non-null  float64
 2   investment_income         10000 non-null  float64
 3   savings_account_balance   10000 non-null  float64
 4   checking_account_balance  10000 non-null  float64
 5   credit_card_balance       10000 non-null  float64
 6   other_loans_balance       10000 non-null  float64
 7   loan_amount_requested     10000 non-null  float64
 8   installment               10000 non-null  float64
 9   loan_date                 10000 non-null  object 
 10  loan_purpose              10000 non-null  object 
 11  state                     10000 non-null  object 
 12  credit_score              10000 non-null  float64
 13  loan_amount               10000 non-null  float64
 14  interes

In [65]:
dfloan_purpose = pd.get_dummies(dfmodel['loan_purpose'],drop_first=True).astype(int)

In [66]:
dfstate = pd.get_dummies(dfmodel['state'],drop_first=True).astype(int)

In [67]:
dfemployment_status = pd.get_dummies(dfmodel['employment_status'],drop_first=True).astype(int)

In [68]:
dfmarital_status = pd.get_dummies(dfmodel['marital_status'],drop_first=True).astype(int)

In [69]:
dfeducation = pd.get_dummies(dfmodel['education_level'],drop_first=True).astype(int)

In [70]:
dfresidential = pd.get_dummies(dfmodel['residential_status'],drop_first=True).astype(int)

In [71]:
dfmodel['loan_date'] = pd.to_datetime(dfmodel['loan_date'])

In [72]:
dfmodel = dfmodel.drop(columns={'loan_purpose','state','employment_status','marital_status','education_level','residential_status'})

In [73]:
dfmodel = pd.concat([dfmodel,dfloan_purpose,dfstate,dfemployment_status,dfmarital_status,dfeducation,dfresidential],axis=1)

In [74]:
col_default = dfmodel['default']

In [75]:
dfmodel = dfmodel.drop(columns={'default'})

In [76]:
dfmodel.insert(35,'default',col_default)

In [77]:
dfmodel

,age,mothly_income,investment_income,savings_account_balance,checking_account_balance,credit_card_balance,other_loans_balance,loan_amount_requested,installment,loan_date,...,Self-Employed,Unemployed,Married,Single,High School,Masters,PhD,Owned,Rented,default
0,56,14300.660,4970.31,14876.86,42372.89,569.14,8652.71,304306.950,13608.01,2020-01-01,...,0,0,0,1,0,0,0,0,1,0
1,69,28491.125,2686.78,85453.24,1353.43,14322.42,23203.16,212348.280,37272.03,2020-01-02,...,0,0,0,1,1,0,0,0,1,1
2,46,6003.690,3550.55,61857.79,10458.77,18613.66,5507.74,87392.494,4694.31,2020-01-03,...,0,0,0,1,0,0,0,1,0,0
3,32,111023.770,281.05,60944.98,44085.33,12644.98,32651.83,234451.610,20010.34,2020-01-04,...,0,0,0,1,0,0,0,1,0,0
4,60,5372.160,3158.68,92134.06,2955.01,14530.57,8461.44,64639.792,5662.19,2020-01-05,...,0,0,1,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,55,87393.670,2402.80,84510.16,22302.94,11777.10,24176.95,668416.830,54435.47,2047-05-14,...,0,0,1,0,0,0,0,0,0,0
9996,51,77018.110,3214.71,28121.09,30456.43,16756.43,6306.73,427142.340,32710.59,2047-05-15,...,1,0,0,0,0,1,0,0,0,0
9997,57,15254.590,3970.41,43204.72,17437.45,13350.67,45261.26,251593.510,10721.52,2047-05-16,...,0,0,1,0,1,0,0,1,0,0
9998,64,18740.570,4476.58,56446.29,36884.46,6084.79,18731.09,5143.840,599.49,2047-05-17,...,1,0,1,0,0,1,0,0,1,0


* Let's see how it goes

In [79]:
X = dfmodel.drop(columns={'default'})

In [80]:
X['loan_date']=pd.to_numeric(X['loan_date'])

In [163]:
y=dfmodel['default'].reset_index()

In [167]:
y.drop(columns={'index'},inplace=True)

In [175]:
y.default

0       0
1       1
2       0
3       0
4       0
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: default, Length: 10000, dtype: int64

In [179]:
from sklearn.model_selection import train_test_split

In [192]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.15, 
    random_state=36,
    stratify=y
   )
print('Treino',round(y_train.mean(),4))
print('Teste',round(y_test.mean(),4))

Treino default    0.1765
dtype: float64
Teste default    0.1767
dtype: float64



* Looks like it tested well


In [195]:
df['default'].value_counts(normalize=True).mul(100)

default
0    82.35
1    17.65
Name: proportion, dtype: float64

In [197]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [199]:
print("\nShape dos dados:")
print(f"X_train: {X_train.shape} | y_train: {y_train.shape}")
print(f"X_test:  {X_test.shape}  | y_test:  {y_test.shape}")

print("\nExemplo de X_train:")
print(X_train.head(3))


Shape dos dados:
X_train: (8500, 35) | y_train: (8500, 1)
X_test:  (1500, 35)  | y_test:  (1500, 1)

Exemplo de X_train:
      age  mothly_income  investment_income  savings_account_balance  \
4256   30      97400.485            4297.84                 64170.15   
885    51      15158.470            4580.94                 59884.99   
4386   29       9082.010            4120.52                 41220.64   

      checking_account_balance  credit_card_balance  other_loans_balance  \
4256                  18543.12             12198.64             23742.81   
885                   15378.08             15706.20              1551.15   
4386                  28774.88              3075.11             45307.90   

      loan_amount_requested  installment            loan_date  ...  TX  \
4256              340009.08     39538.01  1945555200000000000  ...   0   
885                 9717.28      1650.01  1654300800000000000  ...   0   
4386              130291.71      6436.85  1956787200000000000 

In [201]:
from xgboost import XGBClassifier
from sklearn.metrics import classification_report

# Treinar modelo
model = XGBClassifier(
    scale_pos_weight=(len(y_train.default) - sum(y_train.default)) / sum(y_train.default),  # Balanceamento
    eval_metric='aucpr',
    random_state=42
)
model.fit(X_train_scaled, y_train)

# Avaliar
y_pred = model.predict(X_test_scaled)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.99      0.99      1235
           1       0.94      0.92      0.93       265

    accuracy                           0.98      1500
   macro avg       0.96      0.95      0.96      1500
weighted avg       0.98      0.98      0.98      1500



## Ok! We got the precision right! 94% is great 